numba/core/typing/new_builtins.py

```diff
@@ -127,53 +131,506 @@ def generic(self, args, kws):
+    if self.bitwidth == 64:
+        def impl(self):
+            return types.py_complex(self)
+    else:
+        def impl(self):
+            return NotImplemented
+    return impl
+
+
+@overload_method(types.NumPyComplex, "__complex__")
+def np_float64__complex__(self):
+    if self.bitwidth == 128:
+        def impl(self):
+            return types.py_complex(self)
+    else:
+        def impl(self):
+            return NotImplemented
+    return impl
+
+
+@overload_method(types.PythonBoolean, "__add__")
+@overload_method(types.PythonBoolean, "__radd__")
+def py_bool__add__(self, other):
+    def impl(self, other):
+        if isinstance(other, bool) and hasattr(other, "__bool__"):
+            asbool = other.__bool__()
+            if asbool is NotImplemented:
+                return NotImplemented
+            else:
+                return bool_add_bool(self, asbool)
+        elif isinstance(other, int) and hasattr(self, "__int__"):
+            asint = self.__int__()
+            if asint is NotImplemented:
+                return NotImplemented
+            else:
+                return int_add_int(asint, other)
+        else:
+            return NotImplemented
+    return impl
+
+
+@overload_method(types.PythonInteger, "__add__")
+@overload_method(types.PythonInteger, "__radd__")
+def py_int__add__(self, other):
+    # evaluate type
+    def impl(self, other):
+        if isinstance(other, (bool, int)) and hasattr(other, "__int__"):
+            asint = other.__int__()
+            if asint is NotImplemented:
+                return NotImplemented
+            else:
+                return int_add_int(self, asint)
+        else:
+            return NotImplemented
+    return impl
+
+
+@overload_method(types.PythonFloat, "__add__")
+@overload_method(types.PythonFloat, "__radd__")
+def py_float__add__(self, other):
+    # evaluate type
+    def impl(self, other):
+        if hasattr(other, "__float__"):
+            asfloat = other.__float__()
+            if asfloat is NotImplemented:
+                return NotImplemented
+            else:
+                return float_add_float(self, asfloat)
+        else:
+            return NotImplemented
+    return impl
+
+
+@overload_method(types.PythonComplex, "__add__")
+@overload_method(types.PythonComplex, "__radd__")
+def py_complex__add__(self, other):
+        # evaluate type
+        def impl(self, other):
+            if hasattr(other, "__float__"):
+                as_float = other.__float__()
+                if as_float is NotImplemented:
+                    return NotImplemented
+                else:
+                    return complex_add_complex(self, complex(as_float))
+            elif hasattr(other, "__complex__"):
+                as_complex = other.__complex__()
+                if as_complex is NotImplemented:
+                    return NotImplemented
+                else:
+                    return complex_add_complex(self, as_complex)
```

Comment:
see https://github.com/python/cpython/blob/9621a7d0170bf1ec48bcfc35825007cdf75265ea/Objects/complexobject.c#L505-L513

* don't use `__float__`, `__complex__`
* hardcode to only cast instance of `int` and `float`.